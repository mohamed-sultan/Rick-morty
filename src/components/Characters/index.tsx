import { styled } from "@mui/material/styles";
import useMediaQuery from "@mui/material/useMediaQuery";
import { useTheme } from "@mui/material/styles";
import Box from "@mui/material/Box";

import Header from "../Header";
import Footer from "../Footer";
import Card from "../Card";
import SideBar from "../sideBar";
import { characterProps, charactersProps } from "../../interface";

const Wrapper = styled("div")({
  display: "grid",
  gridTemplateColumns: "1fr 1fr 1fr",
  gridColumnGap: "1rem",
  margin: " 5rem 1rem 1rem 1rem",
  alignSelf: "flex-end",
  "@media screen and (max-width: 785px)": {
    display: "unset",
    marginBottom: "15rem",
  },

  "@media screen and (max-width: 1190px)": {
    gridTemplateColumns: "1fr 1fr",
    marginBottom: "15rem",
  },
});

const CharactersHeadLine = styled("h1")(({ theme }) => ({
  color: theme.palette.primary?.black,
  textAlign: "center",
  marginTop: "70px",
  "@media screen and (max-width: 785px)": {
    marginTop: "10rem",
  },
}));

const desktopStyle = {
  display: "flex",
  flexDirection: "row",
  justifyContent: "space-between",
  backgroundColor: "green",
};
const mainStyles = {
  position: "absolute",
  left: "300px",
  top: "100px",
  right: "0px",
  zIndex: 1,
};
const Characters = ({ characters }: charactersProps) => {
  const theme = useTheme();
  const matches = useMediaQuery(theme.breakpoints.up("lg"));

  return (
    <div>
      <Box data-testid="element" sx={!matches ? {} : desktopStyle}>
        <SideBar />
        <div>
          <Header />
          <Box sx={!matches ? {} : mainStyles}>
            <CharactersHeadLine>Rick and Morty characters</CharactersHeadLine>
            <Wrapper>
              {characters?.results.map(
                (character: characterProps | any, i: number) => (
                  <Card character={character} key={i} />
                )
              )}
            </Wrapper>
            <Footer />
          </Box>
        </div>
      </Box>
    </div>
  );
};

export default Characters;
