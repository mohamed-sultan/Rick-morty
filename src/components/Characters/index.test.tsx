import TestRenderer from "react-test-renderer";
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import Characters from "./index";
import mediaQuery from "css-mediaquery";

import { Provider } from "react-redux";
import ConfigureStore from "redux-mock-store";

function createMatchMedia(width: number) {
  return (query: string) => ({
    matches: mediaQuery.match(query, {
      width,
    }),
    addListener: () => {},
    removeListener: () => {},
  });
}

jest.mock("next/router", () => ({
  useRouter: jest.fn(),
}));
describe("Characters snapshot testing", () => {
  const mockStore = ConfigureStore([]);
  const store = mockStore({ ...{ dark: { dark: true } } });
  const data = [
    {
      name: "name",
      image: "/hdflix.png",
      status: "",
      species: "",
      location: { name: "" },
    },
  ];
  const Component = ({ result = [] }) => (
    <Provider store={store}>
      <Characters
        characters={{
          results: result,
        }}
      />
    </Provider>
  );
  it("renders correctly", () => {
    const tree = TestRenderer.create(<Component />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("renders correctly with data", () => {
    const tree = TestRenderer.create(<Component result={data} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("show some flex style for large devices", async () => {
    window.matchMedia = createMatchMedia(1000000);
    const { getByTestId } = render(<Component result={data} />);
    const isVisible = getByTestId("element");
    expect(isVisible).toHaveStyle(`display: flex`);
  });
  it("show some flex style for small devices", async () => {
    window.matchMedia = createMatchMedia(1);
    const { getByTestId } = render(<Component result={data} />);
    const isVisible = getByTestId("element");
    expect(isVisible).not.toHaveStyle(`display: flex`);
  });
});
