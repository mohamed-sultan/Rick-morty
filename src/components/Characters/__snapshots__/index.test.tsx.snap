// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Characters snapshot testing renders correctly 1`] = `
<div>
  <div
    className="MuiBox-root css-0"
    data-testid="element"
  >
    <div>
      <div
        className="css-1e0kmke"
      >
        <span
          className="css-1nedsoq"
        >
          Rick and Morty 
        </span>
        <button
          aria-label="dark-light-icon"
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-9sdrmz-MuiSvgIcon-root"
            data-testid="BedtimeIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12.34 2.02C6.59 1.82 2 6.42 2 12c0 5.52 4.48 10 10 10 3.71 0 6.93-2.02 8.66-5.02-7.51-.25-12.09-8.43-8.32-14.96z"
            />
          </svg>
        </button>
      </div>
      <main
        style={{}}
      >
        <h1
          className="css-jh9j72"
        >
          Rick and Morty characters
        </h1>
        <div
          className="css-jlwmml"
        />
        <footer
          className="css-1ezr195"
        >
          <a
            className="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineNone css-1cmksfb-MuiTypography-root-MuiLink-root"
            href="/2"
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onMouseEnter={[Function]}
            onTouchStart={[Function]}
          >
            <button
              className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-qt782h-MuiButtonBase-root-MuiButton-root"
              disabled={false}
              onBlur={[Function]}
              onContextMenu={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              Next Page
            </button>
          </a>
        </footer>
      </main>
    </div>
  </div>
</div>
`;

exports[`Characters snapshot testing renders correctly with data 1`] = `
<div>
  <div
    className="MuiBox-root css-0"
    data-testid="element"
  >
    <div>
      <div
        className="css-1e0kmke"
      >
        <span
          className="css-1nedsoq"
        >
          Rick and Morty 
        </span>
        <button
          aria-label="dark-light-icon"
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-9sdrmz-MuiSvgIcon-root"
            data-testid="BedtimeIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12.34 2.02C6.59 1.82 2 6.42 2 12c0 5.52 4.48 10 10 10 3.71 0 6.93-2.02 8.66-5.02-7.51-.25-12.09-8.43-8.32-14.96z"
            />
          </svg>
          <span
            className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <main
        style={{}}
      >
        <h1
          className="css-jh9j72"
        >
          Rick and Morty characters
        </h1>
        <div
          className="css-jlwmml"
        >
          <div
            className="css-19i246o"
          >
            <img
              alt="character"
              className="css-y3148r"
              data-nimg="1"
              decoding="async"
              height={300}
              onError={[Function]}
              onLoad={[Function]}
              src="/_next/image?url=%2Fhdflix.png&w=640&q=75"
              srcSet="/_next/image?url=%2Fhdflix.png&w=384&q=75 1x, /_next/image?url=%2Fhdflix.png&w=640&q=75 2x"
              style={
                {
                  "color": "transparent",
                }
              }
              width={300}
            />
            <div>
              <div
                className="css-qs9xs7"
              >
                name
              </div>
              <div
                className="css-xu5fs5"
              >
                 - 
              </div>
              <div
                className="css-1msjh1x"
              />
            </div>
          </div>
        </div>
        <footer
          className="css-1ezr195"
        >
          <a
            className="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineNone css-1cmksfb-MuiTypography-root-MuiLink-root"
            href="/2"
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onMouseEnter={[Function]}
            onTouchStart={[Function]}
          >
            <button
              className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-qt782h-MuiButtonBase-root-MuiButton-root"
              disabled={false}
              onBlur={[Function]}
              onContextMenu={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              Next Page
            </button>
          </a>
        </footer>
      </main>
    </div>
  </div>
</div>
`;
